ci/docker/debian-go-cgo.dockerfile

```diff
@@ -0,0 +1,32 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+ARG base
+FROM ${base}
+
+ENV DEBIAN_FRONTEND noninteractive
+
+# install libarrow-dev to link against with CGO
+RUN apt-get update -y -q && \
+    apt-get install -y -q --no-install-recommends ca-certificates lsb-release wget && \
+    wget https://apache.jfrog.io/artifactory/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb && \
+    apt-get install -y -q --no-install-recommends ./apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb && \
```

Comment:
This uses the latest released Apache Arrow C++ instead of the master of Apache Arrow C++.
Is this intentional?

Reply:
At the moment it was intentional because I didn't want to go through setting up building it with this test and for now I didn't expect there to be cases where changes are being made to both. The drawback of course being that this wouldn't catch a change in the C++ that breaks an API that is being used in the .cc files here though given the desire to have as small a binding as possible, I thought it was acceptable for now.

In order to make this safer, if desired, I could change this to actually build master as part of this like is done in the python tests (note also that `go-cgo-python` dockerfile also uses the latest pyarrow --only-binary for it's test as opposed to building it there)

Reply:
I see.

Reply:
Do you disagree @kou and think I should update it to have it build the c++ off master first and then link against that?

Reply:
I think that it's better that we have at least one CI job that uses master Apache Arrow C++ but I don't disagree with this for now.
We can do it after we face the API breakage problem.